File: check_glb.dart
import 'dart:io';
import 'dart:convert';
import 'dart:typed_data';

void main(List<String> args) async {
  if (args.isEmpty) {
    print('Provide file path');
    return;
  }
  
  final file = File(args[0]);
  if (!await file.exists()) {
    print('File not found: ${args[0]}');
    return;
  }
  
  try {
    final bytes = await file.readAsBytes();
    final byteData = ByteData.view(bytes.buffer);
    
    final magic = String.fromCharCodes(bytes.sublist(0, 4));
    if (magic != 'glTF') {
      print('Not glTF');
      return;
    }
    
    final chunk0Len = byteData.getUint32(12, Endian.little);
    final chunk0Type = String.fromCharCodes(bytes.sublist(16, 20));
    
    if (chunk0Type != 'JSON') {
      print('First chunk is not JSON: $chunk0Type');
      return;
    }
    
    final jsonStr = utf8.decode(bytes.sublist(20, 20 + chunk0Len));
    final gltf = jsonDecode(jsonStr);
    
    final hasAnimations = gltf.containsKey('animations');
    final hasSkins = gltf.containsKey('skins');
    
    print('--- File: ${args[0]} ---');
    print('Has Rig (Skins): $hasSkins');
    if (hasSkins) {
      print('Number of skins: ${(gltf["skins"] as List).length}');
    }
    print('Has Animations: $hasAnimations');
    if (hasAnimations) {
      final anims = gltf['animations'] as List;
      print('Number of animations: ${anims.length}');
      for (var i = 0; i < anims.length; i++) {
        print('  Animation $i: ${anims[i]["name"] ?? "Unnamed"}');
      }
    }
    print('Total Nodes: ${(gltf["nodes"] as List?)?.length ?? 0}');
    print('Total Meshes: ${(gltf["meshes"] as List?)?.length ?? 0}');
    print('---------------------------------');
  } catch (e) {
    print('Error: $e');
  }
}

File: check_glb.py
import struct
import json
import sys

def check_glb(filename):
    try:
        with open(filename, 'rb') as f:
            magic = f.read(4)
            if magic != b'glTF':
                print(f"Not a valid GLB file: {filename}")
                return
            version = struct.unpack('<I', f.read(4))[0]
            length = struct.unpack('<I', f.read(4))[0]
            
            chunk0_length = struct.unpack('<I', f.read(4))[0]
            chunk0_type = f.read(4)
            if chunk0_type != b'JSON':
                print("First chunk is not JSON.")
                return
                
            json_data = f.read(chunk0_length).decode('utf-8')
            gltf = json.loads(json_data)
            
            has_animations = 'animations' in gltf
            has_skins = 'skins' in gltf
            
            print(f"--- File: {filename} ---")
            print(f"Has Rig (Skins): {has_skins}")
            if has_skins:
                print(f"Number of skins: {len(gltf['skins'])}")
                
            print(f"Has Animations: {has_animations}")
            if has_animations:
                print(f"Number of animations: {len(gltf['animations'])}")
                for i, anim in enumerate(gltf['animations']):
                    print(f"  Animation {i}: {anim.get('name', 'Unnamed')}")
                    
            print(f"Total Nodes (Bones/Meshes): {len(gltf.get('nodes', []))}")
            print(f"Total Meshes: {len(gltf.get('meshes', []))}")
            print("---------------------------------")
    except Exception as e:
        print(f"Error reading {filename}: {e}")

if len(sys.argv) > 1:
    check_glb(sys.argv[1])
else:
    print("Please provide a file path.")
